drivers/kscan/kscan_cap1203.c

```diff
@@ -0,0 +1,256 @@
+
+	k_work_submit(&data->work);
+}
+#endif
+
+static int cap1203_configure(const struct device *dev,
+			     kscan_callback_t callback)
+{
+	struct cap1203_data *data = dev->data;
+	const struct cap1203_config *config = dev->config;
+	int r;
+
+	if (!callback) {
+		LOG_ERR("Invalid callback (NULL)");
+		return -EINVAL;
+	}
+
+	data->callback = callback;
+
+#ifdef CONFIG_KSCAN_CAP1203_INTERRUPT
+	/* Clear pending interrupt */
+	r = cap1203_clear_interrupt(&config->i2c);
+	if (r < 0) {
+		return r;
+	}
+
+	r = cap1203_enable_interrupt(&config->i2c, INTERRUPT_ENABLE);
+#else
+	r = cap1203_enable_interrupt(&config->i2c, INTERRUPT_DISABLE);
+#endif
+	if (r < 0) {
+		LOG_ERR("Could not configure interrupt");
+		return r;
+	}
+
+	return 0;
+}
+
+static int cap1203_enable_callback(const struct device *dev)
+{
+	struct cap1203_data *data = dev->data;
+
+#ifdef CONFIG_KSCAN_CAP1203_INTERRUPT
+	const struct cap1203_config *config = dev->config;
+
+	gpio_add_callback(config->int_gpio.port, &data->int_gpio_cb);
+#else
+	k_timer_start(&data->timer, K_MSEC(CONFIG_KSCAN_CAP1203_PERIOD),
+		      K_MSEC(CONFIG_KSCAN_CAP1203_PERIOD));
+#endif
+
+	return 0;
+}
+
+static int cap1203_disable_callback(const struct device *dev)
+{
+	struct cap1203_data *data = dev->data;
+
+#ifdef CONFIG_KSCAN_CAP1203_INTERRUPT
+	const struct cap1203_config *config = dev->config;
+
+	gpio_remove_callback(config->int_gpio.port, &data->int_gpio_cb);
+#else
+	k_timer_stop(&data->timer);
+#endif
+
+	return 0;
+}
+
+static int cap1203_init(const struct device *dev)
+{
+	const struct cap1203_config *config = dev->config;
+	struct cap1203_data *data = dev->data;
+
+	if (!device_is_ready(config->i2c.bus)) {
+		LOG_ERR("I2C controller device not ready");
+		return -ENODEV;
+	}
+
+	data->dev = dev;
+
+	k_work_init(&data->work, cap1203_work_handler);
+
+#ifdef CONFIG_KSCAN_CAP1203_INTERRUPT
+	int r;
+
+	if (!device_is_ready(config->int_gpio.port)) {
+		LOG_ERR("Interrupt GPIO controller device not ready");
+		return -ENODEV;
+	}
+
+	r = gpio_pin_configure_dt(&config->int_gpio, GPIO_INPUT);
+	if (r < 0) {
+		LOG_ERR("Could not confighure interrupt GPIO pin");
+		return r;
+	}
+
+	r = gpio_pin_interrupt_configure_dt(&config->int_gpio,
+					   GPIO_INT_EDGE_TO_ACTIVE);
+	if (r < 0) {
+		LOG_ERR("Could not configure interrupt GPIO interrupt");
+		return r;
+	}
+
+	gpio_init_callback(&data->int_gpio_cb, cap1203_isr_handler,
+			   BIT(config->int_gpio.pin));
+
+#else
+	k_timer_init(&data->timer, cap1203_timer_handler, NULL);
+#endif
+
+	return 0;
+}
+
+static const struct kscan_driver_api cap1203_driver_api = {
+	.config = cap1203_configure,
+	.enable_callback = cap1203_enable_callback,
+	.disable_callback = cap1203_disable_callback,
+};
+
+#define CAP1203_INIT(index)							\
+	static const struct cap1203_config cap1203_config_##index = {		\
+		.i2c = I2C_DT_SPEC_INST_GET(index),				\
+		IF_ENABLED(CONFIG_KSCAN_CAP1203_INTERRUPT,			\
+		(.int_gpio = GPIO_DT_SPEC_INST_GET(index, int_gpios),))		\
```

Comment:
There is a bug here. If there are 2 nodes, and one has `int-gpios`, but the other does not, and CONFIG_KSCAN_CAP1203_INTERRUPT=y, I think this should be using GPIO_DT_SPEC_INST_GET_OR() instead, to set a `NULL` port value in the spec. Then you can just skip devices without interrupt GPIOs, right?

I didn't review in detail though, so not going to block or approve for now.

Reply:
Yes, that was a big vague, sorry. I mean something like this here:

```c
.int_gpio = GPIO_DT_SPEC_INST_GET(index, int_gpios, {0}),
```

Then, in the initialization routine:

```c
if (int_gpio.port != NULL) { do_something_with(&int_gpio); }
```

I.e. we skip the block of code that does something with the interrupt GPIO if there isn't one to do anything with.

> I don't know if this is a bug, but if so, this driver is implemented by imitating [drivers/kscan/kscan_ft3506.c](https://github.com/zephyrproject-rtos/zephyr/blob/70c48417c7a0d7db139accc2368bec7f4aa9c762/drivers/kscan/kscan_ft5336.c#L231-L232), so there may be a bug in kscan_ft5336 as well.

I'm starting to notice this pattern in other reviews as well. From the devicetree side of things, the property is not required, so I think that in the general case, drivers should generally be able to handle devices without such GPIOs even when interrupt support is enabled in software. But if there is something hardware-specific I am missing here, please let me know. Thanks.

Reply:
Thanks, I understand it.

In the current implementation, only interrupt mode or polling mode generates input events from the touch sensor, and this can be switched with Kconfig. Therefore, if interrupt mode is set and there is no int-gpio, the device will not be working.

I think it's okay to switch between using interrupt mode and polling mode depending on the presence or absence of int-gpio, so I'll change it. (However, I think the footprint will be larger.)